lib/routes/account.js

```diff
@@ -333,6 +348,196 @@ module.exports = function (
+                      email: email,
+                      emailCode: crypto.randomBytes(16),
+                      emailVerified: true,
+                      kA: kA,
+                      wrapWrapKb: crypto.randomBytes(32),
+                      accountResetToken: null,
+                      passwordForgotToken: null,
+                      authSalt: authSalt,
+                      verifierVersion: 0,
+                      verifyHash: crypto.randomBytes(32),
+                      openId: id,
+                      verifierSetAt: Date.now(),
+                      locale: locale
+                    }
+                  )
+                }
+              )
+              .then(
+                function (account) {
+                  return db.createSessionToken(
+                    {
+                      uid: account.uid,
+                      email: account.email,
+                      emailCode: account.emailCode,
+                      emailVerified: true,
+                      verifierSetAt: account.verifierSetAt
+                    }
+                  )
+                  .then(
+                    function (sessionToken) {
+                      return db.createKeyFetchToken(
+                        {
+                          uid: account.uid,
+                          kA: account.kA,
+                          // wrapKb is undefined without a password
+                          // wrapWrapKb has the properties we need for this
+                          // value; Its stable, random, and will change on
+                          // account reset.
+                          wrapKb: account.wrapWrapKb,
+                          emailVerified: true
+                        }
+                      )
+                      .then(
+                        function (keyFetchToken) {
+                          return {
+                            sessionToken: sessionToken,
+                            keyFetchToken: keyFetchToken,
+                            unwrapBKey: butil.xorBuffers(
+                              account.kA,
+                              account.wrapWrapKb
+                            )
+                          }
+                        }
+                      )
+                    }
+                  )
+                  .then(
+                    function (tokens) {
+                      reply.redirect(
+                        config.contentServer.url + '/openid?' +
+                        qs.stringify(
+                          {
+                            uid: tokens.sessionToken.uid.toString('hex'),
+                            email: account.email,
+                            session: tokens.sessionToken.data.toString('hex'),
+                            key: tokens.keyFetchToken ?
+                              tokens.keyFetchToken.data.toString('hex')
+                              : undefined,
+                            unwrap: tokens.unwrapBKey.toString('hex'),
```

Comment:
I'm not thrilled about these tokens and key material being in the URL as query parameters, with the attendant risks of leakage through referer logs and the like.  While they don't need to be treated with the same reverence as `kB`, they're still secret data and we should try our best to protect them.

Two suggestions, in increasing order of complexity:

1. Pass then in the hash fragment rather than query parameter, e.g. `/openid?uid=blah#session=foo&key=blah`

2. Indirect through a short-lived "session fetching token".  The idea here would be similar to the key-fetch-token, in that we generate a one-time-use token that can be used to retrieve the session data as a JSON blob.  Pass that back to the content-server like `/openid?uid=blah&session=foo`.  Then the content-server page can extract the token and use it to fetch the actual (and possibly even encrypted!) session data from the auth-server as a separate request.

The advantage of (2) is that even if the "session fetching token" gets leaked some how, it's short-lived and single-use so the potential for exposure is much less.

Or if you wanted to get really fancy, we could have content-server provide a DSA public key when it calls /openid/authenticate, and we could encrypt all the session data to that key before passing it back.  Probably too slow and fragile to be worth the incremental increase in secrecy though, so I suspect I'm getting carried away...

I'm also open to arguments that this won't actually be a problem in practice :smile: